hngprojects/hng_boilerplate_nestjs
# feat: sign in with token endpoint

## What does the PR do?

This PR is about requesting and using an OTP to log in. The endpoint returns errors if:

- [x] User is not registered.
- [x] User inputs the wrong OTP.
- [x] Validation errors on the user inputs invalid or wrong details

How should this be manually tested?

1. Enter your email to request for OTP

2. Make a POST request to the `\api\v1\auth\signin-token` endpoint with the following payload:

```json
{
  "email": "tim@example.com",
}
```

3. A 6-digit OTP is then sent to the user's registered email

4. Then, make a POST request to the `\api\v1\auth\verify-signin-token` endpoint with the following payload:

```json
{
  "email": "tim@example.com",
  "token": "123456"
}
```

Successful Response

On a successful testimonial creation, the API should return a 200 OK status code.

```json
    {
      message: 'Sign-in token sent to email',
      status_code: 200,
    }
```

```json
    {
      message: 'Sign-in successful',
      "token": "eyJhbGciOiJI.......",
      "status_code": 200
    }
```

Authentication Error:

When the user is not registered:

```json
{
  "message": "User not found",
  "status_code": 404
}
```

Invalid OTP error

On Invalid OTP error, if the user is putting an invalid OTP or email:

```json
{
  "message": "Invalid token or email",
  "status_code": 401
}
```

Validation:

The response should return a 400 Bad Request status code with appropriate validation error messages if any required field is missing or invalid.

```json
{
  "message": ["email must be a string", "email must be an email"],
  "status_code": "400"
}
```

Task

- [x] Create a POST endpoint POST /api/v1/auth/signin-token/ to request for OTP for registered users.
- [x] Implement server-side logic to handle successful and failed update requests.
- [x] Ensure the registered user receives an OTP in their email.
- [x] The OTP is properly saved in the database by hashing it
- [x] Check that the OTP is valid
- [x] Then the user gets a token after all validations are passed

Reference Issue: [Link to Issue](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/140)
